File: src/main/java/com/example/chatService/redis/OnlineStatusService.java
package com.example.chatService.redis;

import com.example.chatService.dto.OnlineStatusDto;
import com.example.chatService.dto.UserEnterDto;
import lombok.RequiredArgsConstructor;
import org.springframework.data.redis.core.StringRedisTemplate;
import org.springframework.messaging.simp.SimpMessagingTemplate;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;

import java.time.Duration;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class OnlineStatusService {

    private final StringRedisTemplate redisTemplate;
    private final SimpMessagingTemplate messagingTemplate;

    private static final String ONLINE_HASH = "online:users";
    private static final String TTL_KEY_PREFIX = "online:ttl:";
    private static final Duration EXPIRE_MINUTES = Duration.ofMinutes(2);

    /**
     * 유저 온라인 등록
     */
    public void markOnline(UserEnterDto dto) {
        String userKey = dto.getUserId();

        redisTemplate.opsForHash().put(ONLINE_HASH, userKey, dto.getUsername());
        String ttlKey = TTL_KEY_PREFIX + userKey;
        redisTemplate.opsForValue().set(ttlKey, "1", EXPIRE_MINUTES);

        broadcastOnlineUsers();
    }

    public void refreshTTL(String userId) {
        String key = TTL_KEY_PREFIX + userId;
//        System.out.println("TTL 갱신 " + userId);  // debug
        redisTemplate.expire(key, EXPIRE_MINUTES);
    }

    /**
     * 유저 오프라인 처리
     */
    public void markOffline(String userId) {
        redisTemplate.opsForHash().delete(ONLINE_HASH, userId);
        redisTemplate.delete(TTL_KEY_PREFIX + userId);

        broadcastOnlineUsers();
    }

    public Set<OnlineStatusDto> getAllOnlineUsers() {
        Map<Object, Object> entries = redisTemplate.opsForHash().entries(ONLINE_HASH);

        return entries
                .entrySet()
                .stream()
                .map(e -> new OnlineStatusDto((String) e.getKey(), (String) e.getValue(), true))
                .collect(Collectors.toSet());
    }

    public void broadcastOnlineUsers() {
        Set<OnlineStatusDto> set = getAllOnlineUsers();
        messagingTemplate.convertAndSend("/topic/online-users", set);
    }

    @Scheduled(fixedRate = 5000)
    public void cleanExpiredUsersV2() {
        Set<Object> allUsers = redisTemplate.opsForHash().keys(ONLINE_HASH);

        boolean changed = false;

        for (Object userIdObj : allUsers) {
            String userId = userIdObj.toString();
            String ttlKey = TTL_KEY_PREFIX + userId;

            if (!Boolean.TRUE.equals(redisTemplate.hasKey(ttlKey))) {
                redisTemplate.opsForHash().delete(ONLINE_HASH, userId);

                changed = true;
            }
        }

        if (changed) {
            broadcastOnlineUsers();
        }
    }
}

File: src/main/java/com/example/chatService/redis/OnlineStatusService.kt
package com.example.chatService.redis

import com.example.chatService.dto.OnlineStatusDto
import com.example.chatService.dto.UserEnterDto
import org.springframework.data.redis.core.StringRedisTemplate
import org.springframework.messaging.simp.SimpMessagingTemplate
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Service
import java.time.Duration

@Service
class OnlineStatusService(
    private val redisTemplate: StringRedisTemplate,
    private val messagingTemplate: SimpMessagingTemplate,
) {
    fun markOnline(dto: UserEnterDto) {
        val userKey = dto.userId ?: return
        val username = dto.username ?: return
        redisTemplate.opsForHash<String, String>().put(ONLINE_HASH, userKey, username)
        redisTemplate.opsForValue().set("$TTL_KEY_PREFIX$userKey", "1", EXPIRE_MINUTES)
        broadcastOnlineUsers()
    }

    fun refreshTTL(userId: String) {
        redisTemplate.expire("$TTL_KEY_PREFIX$userId", EXPIRE_MINUTES)
    }

    fun markOffline(userId: String) {
        redisTemplate.opsForHash<String, String>().delete(ONLINE_HASH, userId)
        redisTemplate.delete("$TTL_KEY_PREFIX$userId")
        broadcastOnlineUsers()
    }

    fun getAllOnlineUsers(): Set<OnlineStatusDto> {
        val entries = redisTemplate.opsForHash<String, String>().entries(ONLINE_HASH)
        return entries.entries.mapTo(linkedSetOf()) { OnlineStatusDto(it.key, it.value, true) }
    }

    fun broadcastOnlineUsers() {
        messagingTemplate.convertAndSend("/topic/online-users", getAllOnlineUsers())
    }

    @Scheduled(fixedRate = 5000)
    fun cleanExpiredUsersV2() {
        val allUsers = redisTemplate.opsForHash<String, String>().keys(ONLINE_HASH)
        var changed = false

        for (userId in allUsers) {
            if (redisTemplate.hasKey("$TTL_KEY_PREFIX$userId") != true) {
                redisTemplate.opsForHash<String, String>().delete(ONLINE_HASH, userId)
                changed = true
            }
        }

        if (changed) {
            broadcastOnlineUsers()
        }
    }

    companion object {
        private const val ONLINE_HASH = "online:users"
        private const val TTL_KEY_PREFIX = "online:ttl:"
        private val EXPIRE_MINUTES: Duration = Duration.ofMinutes(2)
    }
}
